core/Pos/Util/Util.hs

```diff
@@ -274,6 +275,12 @@ maybeThrow e = maybe (throwM e) pure
 eitherToFail :: (MonadFail m, ToString s) => Either s a -> m a
 eitherToFail = either (fail . toString) pure
 
+-- | Throw exception or return result depending on what is stored in 'Either'
+eitherToThrow
+    :: (MonadThrow m, Exception e)
+    => (s -> e) -> Either s a -> m a
```

Comment:
I think this function should have type `Either e a -> m a`. At least for consistency with `eitherToFail`. If you need to apply a function to the `Left` value, use `first`.